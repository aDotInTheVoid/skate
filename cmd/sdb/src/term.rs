use std::io::{self, Stdout};

use crossterm::event::{DisableMouseCapture, EnableMouseCapture};
use crossterm::execute;
use crossterm::terminal::{enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen};
use tui::backend::CrosstermBackend;
use tui::terminal::CompletedFrame;
use tui::Frame;

// TODO: Buffer stdout
type Backend = CrosstermBackend<Stdout>;

pub struct Terminal {
    inner: tui::Terminal<Backend>,
}

pub fn install_panic() {
    let old_hook = std::panic::take_hook();
    std::panic::set_hook(Box::new(move |info| {
        if let Ok(mut term) = Terminal::new() {
            let _ = term.close();
        }
        old_hook(info);
    }))
}

impl Terminal {
    pub fn new() -> io::Result<Self> {
        // Set up terminal
        // Copied From `ref/tui-rs/examples/crossterm_demo.rs`
        enable_raw_mode()?;
        let mut stdout = io::stdout();
        execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
        let backend = tui::backend::CrosstermBackend::new(stdout);
        let mut terminal = tui::terminal::Terminal::new(backend)?;
        terminal.clear()?;

        Ok(Self { inner: terminal })
    }

    pub fn close(&mut self) -> io::Result<()> {
        crossterm::terminal::disable_raw_mode()?;
        execute!(
            self.inner.backend_mut(),
            LeaveAlternateScreen,
            DisableMouseCapture
        )?;
        self.inner.show_cursor()?;
        Ok(())
    }

    // Forwarding to [`tui::terminal::Terminal::draw`]
    pub fn draw<F>(&mut self, f: F) -> io::Result<CompletedFrame>
    where
        F: FnOnce(&mut Frame<Backend>),
    {
        self.inner.draw(f)
    }
}
