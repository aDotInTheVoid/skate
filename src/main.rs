#![allow(clippy::all)]
mod grammar; // synthesized by LALRPOP

mod ast;
mod exec;

use codespan_reporting::diagnostic::{Diagnostic, Label};
use codespan_reporting::files::SimpleFiles;
use codespan_reporting::term::emit;
use codespan_reporting::term::termcolor::{ColorChoice, StandardStream};
use fs_err as fs;
use lalrpop_util::ParseError;

// TODO: Make this good.
#[cfg(test)]
mod ast_tests;

fn main() -> eyre::Result<()> {
    let has_error = realmain()?;
    if has_error {
        std::process::exit(1)
    }
    Ok(())
}

// Err(_) -> Fail with interpriter code error
// Ok(false) -> Sucess
// Ok(true) -> Error reported, exit(1) but dont fail
// This is because for an error in user code, we dont want print the eyre backtrace through
// interpriter code, as this is a hot path
fn realmain() -> eyre::Result<bool> {
    let progname = std::env::args()
        .nth(1)
        .ok_or_else(|| eyre::eyre!("Useage: skate <program>"))?;

    let mut err_files = SimpleFiles::new();
    let err_writer = StandardStream::stderr(ColorChoice::Auto);
    let err_config = codespan_reporting::term::Config::default();

    let prog = fs::read_to_string(&progname)?;

    let main_file_id = err_files.add(&progname, &prog);

    // Due to lifetime reasons, we cant convert a parse err to an eyre err
    let prog = grammar::ProgramParser::new().parse(&prog);

    let prog = match prog {
        Ok(p) => p,
        Err(e) => {
            match e {
                ParseError::UnrecognizedToken { token, expected } => {
                    let parse_error = Diagnostic::error()
                        .with_message("Unexpected token")
                        .with_labels(vec![Label::primary(main_file_id, token.0..token.2)])
                        .with_notes(
                            expected
                                .iter()
                                .map(|e| format!("Expected: {}", e))
                                .collect(),
                        );

                    emit(
                        &mut err_writer.lock(),
                        &err_config,
                        &err_files,
                        &parse_error,
                    )?;
                }
                ParseError::InvalidToken { location } => {
                    let parse_error = Diagnostic::error()
                        .with_message("Invalid token")
                        .with_labels(vec![Label::primary(main_file_id, location..location + 1)]);

                    emit(
                        &mut err_writer.lock(),
                        &err_config,
                        &err_files,
                        &parse_error,
                    )?;
                }
                // TODO: Use nice reporting for the rest
                _ => eprintln!("{}", e),
            }

            return Ok(true);
        }
    };

    exec::run(prog)?;

    Ok(false)
}
