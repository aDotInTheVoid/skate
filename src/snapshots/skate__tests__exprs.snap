---
source: src/main.rs
expression: f

---
Literal:
  Integer: 1

